### PR TITLE
Fix a very minor race condition in `cargo fix`.

### PR DESCRIPTION
**Pull request by ehuss**
There is a very rare race where a "fix" message was getting posted *after* `Message::Finish`, and thus being dropped without being printed. This is caused by the diagnostic server disconnecting the client before posting `Message::FixDiagnostic`. The solution is to keep the client on the line until after the message is posted.

Fixes some errors in `fixes_missing_ampersand` seen in CI:
https://travis-ci.org/rust-lang/cargo/jobs/474384359
https://travis-ci.org/rust-lang/cargo/jobs/429809800

I also moved the `done` check to the beginning of the loop because every time the server was shut down it was needlessly trying to parse an empty string (and failing).

### CODE REVIEW

File: src/cargo/util/diagnostic_server.rs [hunk 1]
```diff
@@ -258,14 +258,23 @@ impl RustfixDiagnosticServer {
...
-            }
             if done.load(Ordering::SeqCst) {
                 break;
             }
+            let mut client = BufReader::new(client);
+            let mut s = String::new();
+            if let Err(e) = client.read_to_string(&mut s) {
+                warn!("diagnostic server failed to read: {}", e);
+            } else {
+                match serde_json::from_str(&s) {
```
Comment (dwijnand) [inline src/cargo/util/diagnostic_server.rs]: Why the change from `serde_json::from_reader`?  For the more precise warning message?

Comment (ehuss) [reply src/cargo/util/diagnostic_server.rs]: `from_reader` consumes the reader, and the drops it which closes the connection to the client too early.

**Comment by Eh2406** [comment]
Should this have some kind of test? Not that I know how to make one...

File: src/cargo/util/diagnostic_server.rs (hunk 1 above)
Comment (dwijnand) [reply src/cargo/util/diagnostic_server.rs]: Ah right.

**Comment by dwijnand** [comment]
r=me, after the test chat.

**Comment by alexcrichton** [comment]
To make sure I understand the race here, the problem is that we closed the client connection too early, which allow the client (the `cargo fix` process) to continue and finish. This in turn then finished the entire compilation queue, posting a `Message::Finish`. All of that happened before the server thread was able to post `FixDiagnostic`, causing a message to not get printed. The solution here is to block the client thread until we've actually sent the `FixDiagnostic` message along its way into a queue?

If that's all the case I think it's fine to avoid a test here, spurious failures are showing we're already covering this :)

File: src/cargo/util/diagnostic_server.rs [hunk 2]
```diff
@@ -258,14 +258,23 @@ impl RustfixDiagnosticServer {
...
         while let Ok((client, _)) = self.listener.accept() {
-            let client = BufReader::new(client);
-            match serde_json::from_reader(client) {
-                Ok(message) => on_message(message),
-                Err(e) => warn!("invalid diagnostics message: {}", e),
-            }
             if done.load(Ordering::SeqCst) {
                 break;
             }
+            let mut client = BufReader::new(client);
```
Comment (alexcrichton) [inline src/cargo/util/diagnostic_server.rs]: FWIW with `read_to_string` the `BufReader` here is likely no longer necessary

**Comment by ehuss** [comment]
The test is `fixes_missing_ampersand`.  I can't think of a way to force the race to happen naturally, and it's fairly rare.  I ran it for several hours in a loop on my machines, and on CI, and it seems to be fixed.  You can simulate the error with this PR and the following:

```diff
diff --git a/src/cargo/util/diagnostic_server.rs b/src/cargo/util/diagnostic_server.rs
index e7e1aff1..39140504 100644
--- a/src/cargo/util/diagnostic_server.rs
+++ b/src/cargo/util/diagnostic_server.rs
@@ -266,6 +266,8 @@ impl RustfixDiagnosticServer {
             if let Err(e) = client.read_to_string(&mut s) {
                 warn!("diagnostic server failed to read: {}", e);
             } else {
+                drop(client);
+                std::thread::sleep(std::time::Duration::from_millis(1000));
                 match serde_json::from_str(&s) {
                     Ok(message) => on_message(message),
                     Err(e) => warn!("invalid diagnostics message: {}", e),
@@ -274,7 +276,6 @@ impl RustfixDiagnosticServer {
             // The client should be kept alive until after `on_message` is
             // called to ensure that the client doesn't exit too soon (and
             // Message::Finish getting posted before Message::FixDiagnostic).
-            drop(client);
         }
     }
 }
```

and run `cargo test --test testsuite -- fixes_missing_ampersand` which will give a 100% failure rate.

**Comment by ehuss** [comment]
>  I understand the race here

Yes, everything you said is correct.

File: src/cargo/util/diagnostic_server.rs (hunk 2 above)
Comment (ehuss) [reply src/cargo/util/diagnostic_server.rs]: I looked at that for a bit.  It didn't seem too important, but if I understand the code correctly, a bare read_to_string uses a [32-byte buffer](https://github.com/rust-lang/rust/blob/2442823ef572a65092fbc46f6975633f983b50b6/src/libstd/io/mod.rs#L354) whereas BufReader uses an [8k buffer](https://github.com/rust-lang/rust/blob/2442823ef572a65092fbc46f6975633f983b50b6/src/libstd/io/buffered.rs#L71).  Does that sound correct?  Sometimes it's difficult to trace with various different traits involved.  `read_to_end_with_reservation` doesn't appear to grow its read size, which is a little surprising.

Comment (alexcrichton) [reply src/cargo/util/diagnostic_server.rs]: Eh it could go either way I think. `read_to_string` starts with 32 bytes and exponentially increases (IIRC) and also avoids zeroing and does other fancy tricks. Ideally we'd reuse a buffer here but this isn't really performance sensitive anyway so it doesn't matter too much. Whichever you feel like doing is fine by me!

Comment (ehuss) [reply src/cargo/util/diagnostic_server.rs]: Oh, indeed, it does 32,32,64,128,256,... (I'm guessing it's [this](https://github.com/rust-lang/rust/blob/2442823ef572a65092fbc46f6975633f983b50b6/src/liballoc/raw_vec.rs#L428)).  I'm fine with leaving it as-is, saves a few round trips in the common case (large text messages).

Comment (alexcrichton) [reply src/cargo/util/diagnostic_server.rs]: Ah it's actually [this one](https://github.com/rust-lang/rust/blob/2442823ef572a65092fbc46f6975633f983b50b6/src/libstd/io/mod.rs#L344-L389), but it's morally the same thing!

Comment (ehuss) [reply src/cargo/util/diagnostic_server.rs]: Oh, I was just looking at the end of that call chain, because `read_to_end_with_reservation` does not do any obvious doubling (and the behavior does not appear to be documented anywhere).  I'm assuming the call `g.buf.reserve(32)` goes to `RawVec::reserve` which goes to `reserve_internal` with the `Amortized` strategy to `amortized_new_size` which does the doubling (and the `needed_extra_cap` of 32 gets ignored pretty quickly).

Comment (alexcrichton) [reply src/cargo/util/diagnostic_server.rs]: er oops, right!
